packages/router/src/router.tsx

```diff
@@ -1,34 +1,131 @@
   replaceParams,
   matchPath,
-  mapNamedRoutes,
   PageLoader,
   Redirect,
-  LocationContextType,
-  ParamType,
+  useLocation,
+  validatePath,
+  LocationProvider,
 } from './internal'
+import { PrivateContextProvider, usePrivate } from './private-context'
+import {
+  RouterContextProvider,
+  RouterState,
+  useRouterState,
+} from './router-context'
 import { SplashPage } from './splash-page'
+import { flattenAll, isReactElement } from './util'
+
+/**
+ * A more specific interface will be generated by
+ * babel-plugin-redwood-routes-auto-loader when a redwood project is built
+ *
+ * Example:
+ * interface AvailableRoutes {
+ *   home: () => "/"
+ *   test: () => "/test"
+ * }
+ */
+interface AvailableRoutes {
+  [key: string]: (args?: Record<string, string>) => string
+}
+
+const namedRoutes: AvailableRoutes = {}
+
+type PageType =
+  | Spec
+  | React.ComponentType<unknown>
+  | ((props: any) => JSX.Element)
 
 interface RouteProps {
-  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   path?: string
+  page?: PageType
   name?: string
   notfound?: boolean
   redirect?: string
   prerender?: boolean
-  whileLoading?: () => ReactChild | null
+  whileLoading?: () => React.ReactElement | null
 }
 
-const Route: React.VFC<RouteProps> = () => {
-  return null
+const Route: React.VFC<RouteProps> = ({
+  path,
+  page,
+  name,
+  redirect,
+  notfound,
+  whileLoading = () => null,
+}) => {
+  const location = useLocation()
+  const routerState = useRouterState()
+  const { isPrivate, unauthorized, unauthenticated } = usePrivate()
+
+  if (notfound) {
+    // The "notfound" route is handled by <NotFoundChecker>
+    return null
+  }
+
+  if (!path) {
+    throw new Error(`Route "${name}" needs to specify a path`)
+  }
+
+  // Check for issues with the path.
+  validatePath(path)
+
+  const { match, params: pathParams } = matchPath(
+    path,
+    location.pathname,
+    routerState.paramTypes
+  )
+
+  if (!match) {
+    return null
+  }
+
+  if (isPrivate) {
+    const { loading } = routerState.useAuth()
+
+    if (loading) {
+      return whileLoading()
+    }
+
+    if (unauthorized()) {
+      const currentLocation =
+        global.location.pathname + encodeURIComponent(global.location.search)
+
+      return (
+        <Redirect
+          to={`${namedRoutes[unauthenticated]()}?redirectTo=${currentLocation}`}
+        />
+      )
+    }
+  }
+
+  const searchParams = parseSearch(location.search)
+  const allParams = { ...pathParams, ...searchParams }
+
+  if (redirect) {
+    const newPath = replaceParams(redirect, allParams)
+    return <Redirect to={newPath} />
+  }
+
+  if (!page || !name) {
+    throw new Error(
+      "A route that's not a redirect or notfound route needs to specify both a `page` and a `name`"
+    )
+  }
+
+  return (
+    <PageLoader
+      spec={normalizePage(page)}
+      delay={routerState.pageLoadingDelay}
+      params={allParams}
+    />
+  )
 }
 
 interface PrivateProps {
```

Comment:
🤔 
Thinking out loud here... could `anauthenticated` become: `keyof AvailableRoutes | string`?

Reply:
This works
```
unauthenticated: Exclude<keyof AvailableRoutes, number> | string
```
But TS just simplifies that to `string`, so not much of a point in doing it.

Even in a RW app TS project, where babel has had a chance to properly populate AvailableRoutes, it's still just `string`
![image](https://user-images.githubusercontent.com/30793/111612904-3d8be580-87de-11eb-92a2-ed3c1b03cfce.png)

Reply:
Mkay, it was work a shot!